Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -421,16 +431,195 @@ Further information:
+                                        tasksCopy: inout ContiguousArray<UnderlyingTask>,
+                                        tasksCopyBatchSize: Int) -> NIODeadline? {
+        // We expect empty tasksCopy, to put a new batch of tasks into
+        assert(tasksCopy.isEmpty)
+
+        var moreImmediateTasksToConsider = !immediateTasks.isEmpty
+        var moreScheduledTasksToConsider = !scheduledTasks.isEmpty
+
+        guard moreImmediateTasksToConsider || moreScheduledTasksToConsider else {
+            // Reset nextReadyDeadline to nil which means we will do a blocking select.
+            return nil
+        }
+
+        // We only fetch the time one time as this may be expensive and is generally good enough as if we miss anything we will just do a non-blocking select again anyway.
+        let now: NIODeadline = .now()
+        var nextScheduledTaskDeadline = now
+
+        while moreImmediateTasksToConsider || moreScheduledTasksToConsider {
+            // We pick one item from immediateTasks & scheduledTask per iteration of the loop.
+            // This prevents one task queue starving the other.
+            if moreImmediateTasksToConsider, tasksCopy.count < tasksCopyBatchSize, let task = immediateTasks.popFirst() {
+                tasksCopy.append(task)
+            } else {
+                moreImmediateTasksToConsider = false
+            }
+
+            if moreScheduledTasksToConsider, tasksCopy.count < tasksCopyBatchSize, let task = scheduledTasks.peek() {
+                if task.readyTime.readyIn(now) <= .nanoseconds(0) {
+                    scheduledTasks.pop()
+                    tasksCopy.append(.function(task.task))
+                } else {
+                    nextScheduledTaskDeadline = task.readyTime
+                    moreScheduledTasksToConsider = false
+                }
+            } else {
+                moreScheduledTasksToConsider = false
+            }
+        }
+
+        debugOnly {
+            // The asserts are spun off to a separate functions to aid code clarity
+            // and to remove mutable access to certain structures, e.g. `immediateTasks`.
+            Self._popTasksLockedAssertInvariants(
+                immediateTasks: immediateTasks,
+                                        scheduledTasks: scheduledTasks,
+                                        tasksCopy: tasksCopy,
+                                        tasksCopyBatchSize: tasksCopyBatchSize,
+                                        now: now,
+                                        nextScheduledTaskDeadline: nextScheduledTaskDeadline,
+                                        moreImmediateTasksToConsider: moreImmediateTasksToConsider, moreScheduledTasksToConsider: moreScheduledTasksToConsider)
+        }
+
+
+        // nextScheduledTaskDeadline is the overall next deadline, but iff there are no more immediate tasks left.
+        return moreImmediateTasksToConsider ? now : nextScheduledTaskDeadline
```

Comment:
Will we ever hit `now`? We can only get here if both `moreImmediateTasksToConsider` _and_ `moreScheduledTasksToConsider` are both false, which should imply we never hit `now`.

Reply:
Ah, yes. This changed because the batch size checks got moved to each of the `if`-s, and not the loop overall. I think the statement is harmless, but can be simplified. Checking it carefully.

Reply:
The statement was indeed harmless, because while it'd always return the deadline for next scheduled task, this deadline is only innacted _iff_ there are no more tasks to run. In all other cases we have to recheck, because more tasks may get queued up. Still, this deserved a cleanup, with [`now` returned if there are any more immediate tasks left](https://github.com/apple/swift-nio/pull/2645/commits/7555cc5fae698da0eb2f32d5efad956ff186746e?diff=split&w=0#diff-565b354476ebe678da3be159a85e33e0b7630b1e45b62b20efe506210c1d62faR549) and an [associated assert](https://github.com/apple/swift-nio/pull/2645/commits/7555cc5fae698da0eb2f32d5efad956ff186746e?diff=split&w=0#diff-565b354476ebe678da3be159a85e33e0b7630b1e45b62b20efe506210c1d62faR482-R488). Also, I added a test that tries to get [immediate tasks stuck](https://github.com/apple/swift-nio/pull/2645/commits/7555cc5fae698da0eb2f32d5efad956ff186746e?diff=split&w=0#diff-ee2a5074023719eacac1b5187aeb224fb9afb04f9e010b85809c3f3eb9a2b3b4R1801-R1836) (which never reproduced a problem, but useful defence for the future).